### PR TITLE
fix cpu hotplug vcpu numbers calculation

### PR DESCRIPTION
**Pull request by justin-he**
There are two CPU number in q.config.NumVCPUs(from toml file) and in q.qemuConfig.SMP.CPUs (passed to govmm qemu).
In hotplugAddCPUs(), it didn't use it consistant VCPUs number calculation with resizeVCPUs().

### CODE REVIEW

**Comment by devimc** [comment]
@justin-he please rebase your pr to make travis happy

**Comment by justin-he** [comment]
> @justin-he please rebase your pr to make travis happy

Thanks, done, based on April 11th commit da08b3afc9

File: virtcontainers/sandbox.go [hunk 1]
```diff
@@ -1648,9 +1648,11 @@ func (s *Sandbox) updateResources() error {
 		return fmt.Errorf("sandbox config is nil")
 	}
 
-	sandboxVCPUs := s.calculateSandboxCPUs()
 	// Add default vcpus for sandbox
-	sandboxVCPUs += s.hypervisor.hypervisorConfig().NumVCPUs
+	sandboxVCPUs := s.hypervisor.hypervisorConfig().NumVCPUs
+	if s.calculateSandboxCPUs() != 0 {
```
Comment (devimc) [inline virtcontainers/sandbox.go]: I think you should save this in a variable to avoid calling it twice

**Comment by devimc** [comment]
@egernst implemented this function 
cc @jcvenegas

**Comment by jcvenegas** [comment]
hey @justin-he  thanks for the PR, this is something need get more feedback from @kata-containers/architecture-committee.

The way that is implemented today the calculation is to additionally add whatever the value we provide from the config. This will be used for non constrained containers, and this PR still keeps the way to do it in the case of single containers model, but I think it will not add the additional cpus in case of k8s. 

Finally the other reason still include is to add bit more of cpu resources for agent and things running in the VM that are not part of the container.

**Comment by egernst** [comment]
@justin-he PTAL @ https://github.com/kata-containers/documentation/blob/master/design/cpu-constraints.md#default-number-of-virtual-cpus

This explains the current implementation.  It isn't completely intuitive, I hope the document clarifies.  If you think this should be changed, then let's continue the conversation from a design POV.

**Comment by justin-he** [comment]
Ok，I will drop the 2nd patch. What about the 1st one?

**Comment by justin-he** [comment]
Changed the descriptions of this pr after dropping 2nd patch

**Comment by egernst** [comment]
Is it feasible to add a unit test to show the prior failure, and verify this fix?

**Comment by jcvenegas** [comment]
@justin-he @egernst  I think the result is the same it is just to be consistent with the rest of the code?. 
This is more a duplicated information issue, in different structs.

**Comment by justin-he** [comment]
> @justin-he @egernst I think the result is the same it is just to be consistent with the rest of the code?.
> This is more a duplicated information issue, in different structs.

Yes，  result is the same, just to keep  consistent with the rest of the code

**Comment by jodh-intel** [comment]
Any update on this folks?

**Comment by jcvenegas** [comment]
@justin-he hi any update about this related to make the code consistent but using `q.qemuConfig.SMP.CPUs` in other places

**Comment by justin-he** [comment]
> @justin-he hi any update about this related to make the code consistent but using `q.qemuConfig.SMP.CPUs` in other places

OK, I will check it asap

**Comment by justin-he** [comment]
@jcvenegas Please review whether I understand it correctly, thanks

**Comment by raravena80** [comment]
@justin-he @jcvenegas any updates? Thx!

**Comment by grahamwhaley** [comment]
@jcvenegas @devimc - I think this one is pending your re-review from my reading of the history - pls have a peek, thx.

**Comment by caoruidong** [comment]
ping @justin-he

**Comment by justin-he** [comment]
@caoruidong Thanks, I am not familiar community's process. Seems the test result is not available for me to check now, could you please trigger a test again. Sorry for the inconvenience

**Comment by caoruidong** [comment]
OK. /test

**Comment by justin-he** [comment]
@gnawux @caoruidong Seems the CI error is not related to my patch, @Pennyzct had posted a [`PR`](https://github.com/kata-containers/tests/pull/1729) to fix it.

**Comment by jcvenegas** [comment]
@chavafg @GabyCT  seems that  vsocks job is not working `jenkins-ci-fedora-vsocks — Build finished.`.
@justin-he sorry to retrigger test again, I'll merge it as soon it finish again, yeah ARM error is not related with it.

**Comment by chavafg** [comment]
@jcvenegas the name of the job changed to `jenkins-ci-vsocks`. The old one will continue being there as failed

File: virtcontainers/qemu.go [hunk 2]
```diff
@@ -1669,7 +1669,7 @@ func calcHotplugMemMiBSize(mem uint32, memorySectionSizeMB uint32) (uint32, erro
 
 func (q *qemu) resizeVCPUs(reqVCPUs uint32) (currentVCPUs uint32, newVCPUs uint32, err error) {
 
-	currentVCPUs = q.config.NumVCPUs + uint32(len(q.state.HotpluggedVCPUs))
+	currentVCPUs = q.qemuConfig.SMP.CPUs + uint32(len(q.state.HotpluggedVCPUs))
```
Comment (jodh-intel) [inline virtcontainers/qemu.go]: I think this justifies a new unit test to check all the "awkward" corner case values and assert the expected behaviour. So, how about wrapping this logic in a new function to make writing that test easier? Something like:

```go
func (q *qemu) calculateCurrentVCPUs() uint32 {
    // ...
}
```

**Comment by raravena80** [comment]
Ping @jcvenegas @chavafg CI issue?

**Comment by jodh-intel** [comment]
Hi @justin-he - this branch is now conflicted. Please could you rebase and re-push?

**Comment by justin-he** [comment]
> Hi @justin-he - this branch is now conflicted. Please could you rebase and re-push?

Ok, thanks

**Comment by jodh-intel** [comment]
Hi @justin-he - are you sure you force-pushed as it's still showing as conflicted?

**Comment by justin-he** [comment]
> Hi @justin-he - are you sure you force-pushed as it's still showing as conflicted?

No, I push -f and see no conflicts reported

**Comment by jodh-intel** [comment]
@justin-he - odd, if you look at the bottom of this PR, it still shows:

```
This branch has conflicts that must be resolved

Conflicting files

virtcontainers/qemu.go

```

**Comment by justin-he** [comment]
> @justin-he - odd, if you look at the bottom of this PR, it still shows:
> 
> ```
> This branch has conflicts that must be resolved
> 
> Conflicting files
> 
> virtcontainers/qemu.go
> ```
ah, got it, there is a bug in commit 50e263d9 IIUC

File: virtcontainers/qemu.go (hunk 2 above)
Comment (jodh-intel) [reply virtcontainers/qemu.go]: Thanks for resolving the conflict. The area of vcpu calculation has seen a few changes so I do think a unit test would help others understand the logic here. Take a look at the following for the basic template for a table-based approach:

- https://github.com/kata-containers/tests/blob/master/Unit-Test-Advice.md.

Comment (justin-he) [reply virtcontainers/qemu.go]: ok, James, thanks. I will do it later

**Comment by raravena80** [comment]
@justin-he any updates on this PR? Thx!

**Comment by hejianet** [comment]
Hi @raravena80 Ricardo, sorry for the late response. Since it is not an urgent fix, could we close it now?
Sorry if there is any inconvenience

**Comment by raravena80** [comment]
@hejianet Do you mean to close it as in making this PR go through or as in 'won't fix'?

**Comment by hejianet** [comment]
> @hejianet Do you mean to close it as in making this PR go through or as in 'won't fix'?

Yes. sorry about that :(
